### PR TITLE
[RFC] Extract failed test debugger code into its own utiltiy class

### LINKED ISSUE #420: Improve failed test's AssertionError message mapping
As discussed in the Gitter Chat, a few other ideas for improving #418 are:

- extract the third argument (FxRobot) in verifyThat into a DebugUtils class. Then, one could write verifyThat(node, matcher, DebugUtils.informedErrorHandler(thisAsFxRobot)
- in the above DebugUtils class, separate items into functions that can be composed, so that one can quickly add only the things they need to see (e.g. keys pressed, buttons pressed, etc.)
- change the mapping function to `UnaryOperator<String>`, so that the developer can run additional things within the error handler without forgetting to return a new AssertionError (they may return the original one). Then, the error handler code would look like:

````java
try {
    matcher.match(node);
} catch (AssertionError error) {
    throw`new AssertionError(mapper.apply(error.getMessage));
}
````

### PR DESCRIPTION
**Pull request by JordanMartinez**
Resolves #420

### CODE REVIEW

File: subprojects/testfx-core/src/main/java/org/testfx/api/FxAssert.java [hunk 1]
```diff
@@ -38,6 +35,11 @@
 import org.testfx.service.finder.NodeFinder;
 import org.testfx.service.query.NodeQuery;
 
+/**
+ * All TestFX tests should use {@link #verifyThat(Node, Matcher, Function)} when writing tests,
```
Comment (brcolow) [inline subprojects/testfx-core/src/main/java/org/testfx/api/FxAssert.java]: Maybe we should be explicit and say "All internal TestFX tests"

Comment (JordanMartinez) [inline subprojects/testfx-core/src/main/java/org/testfx/api/FxAssert.java]: Why? I really meant all tests should use the `verifyThat(Node, Matcher)`-related assertions because if something goes wrong, one can quickly add an error handler (type now becomes `verifyThat(Node, Matcher, Function)`) that provides additional info to help one debug it. This applies to both TestFX contributors and users of this test library

**Comment by JordanMartinez** [comment]
Before merging, this PR should include tests that prove the code works....

File: subprojects/testfx-core/src/main/java/org/testfx/api/FxAssert.java (hunk 1 above)
Comment (JordanMartinez) [inline subprojects/testfx-core/src/main/java/org/testfx/api/FxAssert.java]: Perhaps I should reword it to explain that better?

File: subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+import static org.testfx.util.DebugUtils.showFiredEvents;
+import static org.testfx.util.DebugUtils.showKeysPressedAtTestFailure;
+import static org.testfx.util.DebugUtils.showMouseButtonsPressedAtTestFailure;
+
+public class DebugUtilsTest extends TestCaseBase {
+
+    private static final String TAB = " ";
+
+    private static final MouseButton MOUSE_BUTTON = MouseButton.PRIMARY;
+    private static final KeyCode KEY = KeyCode.A;
```
Comment (brcolow) [inline subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java]: I don't think that these fields add anything useful. I understand if the intention is to make it clear that it can be any mouse button or key, but I think it complicates reading the test.

File: subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+        ) {
+
+            @Override
+            public boolean matches(Object item) {
+                return !replaceBlankPart(item).contains(text);
+            }
+        };
+    }
+
+    private abstract class RemoveBlankMessageMatcher extends BaseMatcher<String> {
```
Comment (brcolow) [inline subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java]: What exactly does this class do? Having a hard time understanding it.

File: subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+            FxAssert.verifyThat(false, alwaysFail(), errorMessageMapper);
+            throw new AssertionError("Error: verifyThat did not throw an error when it should have.");
+        }
+        catch (AssertionError error) {
+            return error;
+        }
+    }
+
+    private Matcher<Object> alwaysFail() {
+        return new AlwaysFailsMatcher();
```
Comment (brcolow) [inline subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java]: Could this be replaced by returning an anonymous inner class that extends Hamcrest's `IsNot` matcher with `not()` overriden to always return `true`? Would that make things any clearer? Not sure.

File: subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+        AssertionError error = getThrownErrorPostMapper(showFiredEvents(events, TAB));
+
+        // then:
+        assertThat(error.getMessage(), hasAddedTextThatContains("MouseEvent"));
+        assertThat(error.getMessage(), hasAddedTextNotContaining("ScrollEvent"));
+    }
+
+    private AssertionError getThrownErrorPostMapper(Function<StringBuilder, StringBuilder> errorMessageMapper) {
+        try {
+            FxAssert.verifyThat(false, alwaysFail(), errorMessageMapper);
```
Comment (brcolow) [inline subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java]: Can we static import `verifyThat` here?

File: subprojects/testfx-core/src/main/java/org/testfx/util/DebugUtils.java [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+ * {@link org.testfx.api.FxAssert#verifyThat(Node, Matcher)} or its related methods.
+ *
+ * <p>
+ *     Quickly chain things together using {@link #compose(Function[])} or use standard handlers like
+ *     {@link #informedErrorMessage(String, boolean, FxRobot, boolean, boolean)}. Any {@code tab} parameters in methods
+ *     are there to specify what to insert as a tab. This could be a tab or a number of spaces. Default tab
+ *     value is three spaces (e.g. {@code "   "}).
+ * </p>
+ */
+public final class DebugUtils {
```
Comment (brcolow) [inline subprojects/testfx-core/src/main/java/org/testfx/util/DebugUtils.java]: Should we maybe say "are there to specify what text is inserted for spacing purposes. This defaults to blah blah". What I mean is instead of explicitly calling it "tab", maybe rename to `spacing`?

File: subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java (hunk 2 above)
Comment (JordanMartinez) [inline subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java]: If the test's assertion should fail, I need to release the mouse button or key that was pressed. Rather than defining that in two separate places that can change, I define it once here and refer to it twice (1x in test, 1x in cleanup).

File: subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java (hunk 5 above)
Comment (JordanMartinez) [inline subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java]: Sure thing!

File: subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java (hunk 4 above)
Comment (JordanMartinez) [inline subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java]: I'm not familiar with Hamcrest and their matchers, so I ran my own after looking at the ones created in TestFX itself.

It probably would make things clearer if a developer reading through these test knows Hamcrest well.

File: subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java (hunk 3 above)
Comment (JordanMartinez) [inline subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java]: It removes the blank part (i.e. the the quotes part in `"Expected: <expectedValue>\n\tbut: "<actualValue>") from the returned error's message since I can't get rid of that myself but still need a better way to determine whether a test worked or not. 

I don't think it's the best designed class because I initially made it do something else before reworking it a bit. The ">>>message<<<" was necessary for debugging my tests until I replaced newline characters with the "\n" string. I'm not sure if it's needed anymore.

File: subprojects/testfx-core/src/main/java/org/testfx/util/DebugUtils.java (hunk 6 above)
Comment (JordanMartinez) [inline subprojects/testfx-core/src/main/java/org/testfx/util/DebugUtils.java]: Perhaps a better name would be "indent"? 

"Tab" is good in that that is what it's used for, indenting stuff. However, it's bad in that one may assume one needs to pass in "\t". "Spacing" isn't good either, though it fixes the problem with "tab," in that it doesn't communicate the indenting nature of it all. What are we spacing? Space between the current added text and the next one?

File: subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java (hunk 4 above)
Comment (JordanMartinez) [inline subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java]: I looked into it. `IsNot` requires another Matcher to be passed into it. So, I think it's just easier to leave it as is. Besides, it also overrides the `describeMismatch()` to remove the "was" that gets appended to the description.

File: subprojects/testfx-core/src/main/java/org/testfx/util/DebugUtils.java (hunk 6 above)
Comment (brcolow) [inline subprojects/testfx-core/src/main/java/org/testfx/util/DebugUtils.java]: Indent works better IMO

File: subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java (hunk 3 above)
Comment (JordanMartinez) [inline subprojects/testfx-core/src/test/java/org/testfx/util/DebugUtilsTest.java]: I just realized that this isn't actually needed. Why? Because there's a much easier way to remove the blank part by taking any function `f` and calling `f.compose(sb -> new StringBuilder())`

**Comment by brcolow** [comment]
Awesome. Thanks @JordanMartinez
